File: core/src/com/mygdx/game/Screens/PlayScreen.java
package com.mygdx.game.Screens;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Screen;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.maps.MapObject;
import com.badlogic.gdx.maps.objects.RectangleMapObject;
import com.badlogic.gdx.maps.tiled.TiledMap;
import com.badlogic.gdx.maps.tiled.TmxMapLoader;
import com.badlogic.gdx.maps.tiled.renderers.OrthogonalTiledMapRenderer;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.Body;
import com.badlogic.gdx.physics.box2d.BodyDef;
import com.badlogic.gdx.physics.box2d.Box2DDebugRenderer;
import com.badlogic.gdx.physics.box2d.FixtureDef;
import com.badlogic.gdx.physics.box2d.PolygonShape;
import com.badlogic.gdx.physics.box2d.World;
import com.badlogic.gdx.utils.viewport.FitViewport;
import com.badlogic.gdx.utils.viewport.Viewport;
import com.mygdx.game.Main;
import com.mygdx.game.Scenes.HUD;
import com.mygdx.game.Sprites.Player;

public class PlayScreen implements Screen {

    Main game;
    OrthographicCamera camera;
    Viewport viewport;
    HUD hud;

    private Player player;

    private TmxMapLoader mapLoader;
    private TiledMap map;
    private OrthogonalTiledMapRenderer renderer;

    private World world;
    private Box2DDebugRenderer b2dr;

    public PlayScreen(Main game) {
        this.game = game;
        camera = new OrthographicCamera();
        viewport = new FitViewport(Main.V_WIDTH, Main.V_HEIGHT, camera);
        hud = new HUD(game.batch);

        mapLoader = new TmxMapLoader();
        map = mapLoader.load("lv1.tmx");
        renderer = new OrthogonalTiledMapRenderer(map);
        camera.position.set(viewport.getWorldWidth() / 2, viewport.getWorldHeight() / 2, 0);

        world = new World(new Vector2(0, -10), true);
        player = new Player(world);
        b2dr = new Box2DDebugRenderer();

        BodyDef bodyDef = new BodyDef();
        PolygonShape shape = new PolygonShape();
        FixtureDef fixtureDef = new FixtureDef();
        Body body;

        // 2d Box ground
        //  Ground
        for (MapObject object : map.getLayers().get(1).getObjects().getByType(RectangleMapObject.class)) {
            Rectangle rect = ((RectangleMapObject) object).getRectangle();
            bodyDef.type = BodyDef.BodyType.StaticBody;
            bodyDef.position.set(rect.getX() + rect.getWidth() / 2, rect.getY() + rect.getHeight() / 2);
            body = world.createBody(bodyDef);
            shape.setAsBox(rect.getWidth() / 2, rect.getHeight() / 2);
            fixtureDef.shape = shape;
            body.createFixture(fixtureDef);
        }
        //  Bricks
        for (MapObject object : map.getLayers().get(2).getObjects().getByType(RectangleMapObject.class)) {
            Rectangle rect = ((RectangleMapObject) object).getRectangle();
            bodyDef.type = BodyDef.BodyType.StaticBody;
            bodyDef.position.set(rect.getX() + rect.getWidth() / 2, rect.getY() + rect.getHeight() / 2);
            body = world.createBody(bodyDef);
            shape.setAsBox(rect.getWidth() / 2, rect.getHeight() / 2);
            fixtureDef.shape = shape;
            body.createFixture(fixtureDef);
        }

    }

    public void handleInput(float delta) {
//        if (Gdx.input.isTouched()) {
//            camera.position.x += 100 * delta;
//        }
//        if (Gdx.input.isKeyJustPressed(Input.Keys.UP))
//            player.b2Body.applyLinearImpulse(new Vector2(0, 4f), player.b2Body.getWorldCenter(), true);
//
//        if (Gdx.input.isKeyJustPressed(Input.Keys.RIGHT) && player.b2Body.getLinearVelocity().x <= 2)
        if (Gdx.input.isTouched())
            player.b2Body.applyLinearImpulse(new Vector2(0.5f, 0), player.b2Body.getWorldCenter(), true);
//
//        if (Gdx.input.isKeyJustPressed(Input.Keys.LEFT) && player.b2Body.getLinearVelocity().x >= -2)
//            player.b2Body.applyLinearImpulse(new Vector2(0, 0.1f), player.b2Body.getWorldCenter(), true);


    }

    public void update(float delta) {
        handleInput(delta);
        world.step(1 / 60f, 6, 2);
        camera.position.x = player.b2Body.getPosition().x;
        camera.update();
        renderer.setView(camera);
    }

    @Override
    public void show() {

    }

    @Override
    public void render(float delta) {
        update(delta);
        Gdx.gl.glClearColor(0, 0, 0, 1);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

        game.batch.setProjectionMatrix(hud.stage.getCamera().combined);
        hud.stage.draw();

        renderer.render();
        b2dr.render(world, camera.combined);
    }

    @Override
    public void resize(int width, int height) {
        viewport.update(width, height);
    }

    @Override
    public void pause() {

    }

    @Override
    public void resume() {

    }

    @Override
    public void hide() {

    }

    @Override
    public void dispose() {

    }
}

File: core/src/com/mygdx/game/Screens/PlayScreen.kt
package com.mygdx.game.Screens

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Screen
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.maps.tiled.TiledMap
import com.badlogic.gdx.maps.tiled.TmxMapLoader
import com.badlogic.gdx.maps.tiled.renderers.OrthogonalTiledMapRenderer
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.physics.box2d.Box2DDebugRenderer
import com.badlogic.gdx.physics.box2d.World
import com.badlogic.gdx.utils.viewport.FitViewport
import com.badlogic.gdx.utils.viewport.Viewport
import com.mygdx.game.Controller.OnScreenController
import com.mygdx.game.MainGame
import com.mygdx.game.Scenes.Hud
import com.mygdx.game.Sprites.Player
import com.mygdx.game.Tools.B2WorldCreator
import com.mygdx.game.Tools.drawSprite

class PlayScreen(var game: MainGame) : Screen {
    private val mapLoader: TmxMapLoader = TmxMapLoader()
    private val tiledMap: TiledMap
    private val mapRenderer: OrthogonalTiledMapRenderer
    val textureAtlas: TextureAtlas = TextureAtlas("Dino.pack")

    private var camera: OrthographicCamera = OrthographicCamera()
    private var viewport: Viewport

    private val b2dr: Box2DDebugRenderer

    private var hud: Hud
    private val world: World

    private val player: Player
    private val onScreenController: OnScreenController

    init {
        tiledMap = mapLoader.load("lv1.tmx")
        mapRenderer = OrthogonalTiledMapRenderer(tiledMap)

        viewport = FitViewport(MainGame.V_WIDTH.toFloat(), MainGame.V_HEIGHT.toFloat(), camera)
        camera.position[viewport.worldWidth / 2, viewport.worldHeight / 2] = 0f
        b2dr = Box2DDebugRenderer()
        hud = Hud(game.batch)

        world = World(Vector2(0f, -80f), true)
        player = Player(world, this)
        onScreenController = OnScreenController(game)

        B2WorldCreator(world, tiledMap)
    }

    private fun handleInput(delta: Float) {
        if (onScreenController.isLeftPressed || onScreenController.isRightPressed) {
            player.moveHorizontal(onScreenController.isLeftPressed)
        } else if (onScreenController.isUpPressed) {
            player.jump()
        }
    }

    private fun updateWorldStatus(delta: Float) {
        handleInput(delta)
        world.step(1 / 60f, 6, 2)
        // Attach camera to player
        camera.position.x = player.b2Body.position.x

        player.update(delta)
        camera.update()
        // Draw only visible part of the world
        mapRenderer.setView(camera)
    }

    override fun render(delta: Float) {
        updateWorldStatus(delta)
        clearScreen()

//        game.batch.projectionMatrix = hud.stage.camera.combined
        game.batch.projectionMatrix = camera.combined
        game.batch.begin()
        game.batch.drawSprite(player)
        game.batch.end()

        hud.stage.draw()
        mapRenderer.render()
        onScreenController.draw()
        // Debug purposes
        b2dr.render(world, camera.combined)
    }

    override fun resize(width: Int, height: Int) {
        viewport.update(width, height)
        onScreenController.resize(width, height)
    }

    override fun dispose() {
        tiledMap.dispose()
        mapRenderer.dispose()
        world.dispose()
        b2dr.dispose()
        hud.dispose()
    }

    private fun clearScreen() {
        Gdx.gl.glClearColor(0f, 0f, 0f, 1f)
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT)
    }

    override fun pause() {}
    override fun resume() {}
    override fun hide() {}
    override fun show() {}
}